android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedNotificationsEmitter.java

```diff
@@ -0,0 +1,102 @@
+package versioned.host.exp.exponent.modules.universal.notifications;
+
+import android.content.Context;
+import android.os.Bundle;
+import android.util.Log;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Map;
+
+import expo.modules.notifications.notifications.emitting.NotificationsEmitter;
+import expo.modules.notifications.notifications.interfaces.NotificationTrigger;
+import expo.modules.notifications.notifications.model.Notification;
+import expo.modules.notifications.notifications.model.NotificationResponse;
+import expo.modules.notifications.notifications.model.triggers.FirebaseNotificationTrigger;
+import host.exp.exponent.kernel.ExperienceId;
+
+import static versioned.host.exp.exponent.modules.universal.notifications.ScopedNotificationScheduler.EXPERIENCE_ID_KEY;
+import static versioned.host.exp.exponent.modules.universal.notifications.ScopedNotificationScheduler.USER_DATA_KEY;
+
+public class ScopedNotificationsEmitter extends NotificationsEmitter {
+  private ExperienceId mExperienceId;
+
+  public ScopedNotificationsEmitter(Context context, ExperienceId experienceId) {
+    super(context);
+    mExperienceId = experienceId;
+  }
+
+  @Override
+  public void onNotificationReceived(Notification notification) {
+    if (shouldHandleNotification(notification)) {
+      super.onNotificationReceived(notification);
+    }
+  }
+
+  @Override
+  public void onNotificationResponseReceived(NotificationResponse response) {
+    if (shouldHandleNotification(response.getNotification())) {
+      super.onNotificationResponseReceived(response);
+    }
+  }
+
+  private boolean shouldHandleNotification(Notification notification) {
+    NotificationTrigger notificationTrigger = notification.getNotificationRequest().getTrigger();
+    if (notificationTrigger instanceof FirebaseNotificationTrigger) {
+      Map<String, String> data = ((FirebaseNotificationTrigger) notificationTrigger).getRemoteMessage().getData();
+      if (!data.containsKey("experienceId")) {
+        return false;
+      }
+      String experienceIdString = data.get("experienceId");
+      return mExperienceId.get().equals(experienceIdString);
+    } else {
+      JSONObject body = notification.getNotificationRequest().getContent().getBody();
+      if (body == null) {
+        return false;
+      }
+      try {
+        String experienceId = body.getString(EXPERIENCE_ID_KEY);
+        return mExperienceId.get().equals(experienceId);
+      } catch (JSONException e) {
+        Log.w("NotificationsEmitter", String.format("The notification's body should contains '%s' field.", EXPERIENCE_ID_KEY), e);
+      }
+    }
+    return false;
+  }
+
+  @Override
+  protected void emitEvent(String event, Bundle eventData) {
+    super.emitEvent(event, removeExpoData(eventData));
+  }
+
+  private Bundle removeExpoData(Bundle eventData) {
```

Comment:
Huh… although this method is beautiful I've been wondering if we could make it… not be at all. One idea that came to my mind was that there is only a handful of places in which we create instances of `NotificationRequest`:
1. `FirebaseListenerService`, when a push notification is received
2. `ExpoNotificationPresentationModule`, when it… presents the notification
3. `NotificationScheduler` when we create a `NotificationRequest` to be scheduled

What if we created a subclass of `NotificationRequest` which would contain an experience ID to which it should be delivered? We already have subclassing access to all three of those places:
- in Firebase listener service we could infer the `experienceId` from the `RemoteMessage`'s `data`
- in both exported modules we have access to `mExperienceId`

Good news is that:
- we wouldn't have to scope and unscope `data` anywhere
- we could remove the `experienceId` parameter from `data` in the subclass of `FirebaseListenerService`

I don't see any disadvantages yet… do you?

Reply:
That sounds great :D I will apply this ;)